File: papers/miniKanren/ocanren/app/app1.ml
open GT
open OCanren
open OCanren.Std
open Helper

let topLevel x0 x1 x2 =
  let rec appendo y0 y1 y2 =
    fresh (q1 q2 q3 q4 q5 q6)
      ( y0 === Std.List.nil () &&& (y1 === y2)
      ||| ( y0 === Std.( % ) q1 q2
          &&& (y2 === Std.( % ) q1 q3)
          &&& (q2 === Std.List.nil () &&& (y1 === q3) ||| (q2 === Std.( % ) q4 q5 &&& (q3 === Std.( % ) q4 q6) &&& _appendo y1 q5 q6)) ) )
  and _appendo y3 y4 y5 =
    fresh (q1 q2 q3) (y4 === Std.List.nil () &&& (y3 === y5) ||| (y4 === Std.( % ) q1 q2 &&& (y5 === Std.( % ) q1 q3) &&& _appendo y3 q2 q3))
  in
  appendo x0 x1 x2

File: papers/miniKanren/prolog/app/app1.pl
appendo(nil, Y2, Y2).
appendo(cons(Q1, nil), Q3, cons(Q1, Q3)).
appendo(cons(Q1, cons(Q4, Q5)), Y1, cons(Q1, cons(Q4, Q6))) :- _appendo(Y1, Q5, Q6).
_appendo(Y5, nil, Y5).
_appendo(Y3, cons(Q1, Q2), cons(Q1, Q3)) :- _appendo(Y3, Q2, Q3).